From f3a2c2344cab615f2c0bff932a51154f104c3836 Mon Sep 17 00:00:00 2001
From: Alexander Bokovoy <abokovoy@redhat.com>
Date: Wed, 7 Oct 2026 09:33:02 +0300
Subject: [PATCH] KDCKEMInfo: kemct and serverNonce are EXPLICIT tagged

draft-bokovoy-kitten-pkinit-pqc-02 Section 6.3 declares KDCKEMInfo in
the EXPLICIT TAGS KerberosV5-PK-INIT-SPEC module, with no IMPLICIT
keyword on any field, so kemct [1] and serverNonce [4] are EXPLICIT
tagged, like kemAlgorithm, kdfAlgorithm and nonce already are. Encoding
them IMPLICIT made the KDC's KDCKEMInfo unreadable to other draft-02
implementations (and theirs to this client).
---
 src/lib/krb5/asn.1/asn1_k_encode.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/lib/krb5/asn.1/asn1_k_encode.c b/src/lib/krb5/asn.1/asn1_k_encode.c
index 1329bc9f1..896227a91 100644
--- a/src/lib/krb5/asn.1/asn1_k_encode.c
+++ b/src/lib/krb5/asn.1/asn1_k_encode.c
@@ -1409,12 +1409,12 @@ MAKE_DECODER(decode_krb5_pa_pk_as_req_hint, pa_pk_as_req_hint);
 /* KDCKEMInfo (draft-bokovoy-kitten-pkinit-pqc) */
 DEFFIELD(kem_info_0, krb5_kdc_kem_info, kemAlgorithm, 0,
          algorithm_identifier);
-DEFFIELD_IMPLICIT(kem_info_1, krb5_kdc_kem_info, kemct, 1, ostring_data);
+DEFFIELD(kem_info_1, krb5_kdc_kem_info, kemct, 1, ostring_data);
 DEFFIELD(kem_info_2, krb5_kdc_kem_info, kdfAlgorithm, 2,
          algorithm_identifier);
 DEFFIELD(kem_info_3, krb5_kdc_kem_info, nonce, 3, opt_int32);
-DEFFIELD_IMPLICIT(kem_info_4, krb5_kdc_kem_info, serverNonce, 4,
-                  opt_ostring_data);
+DEFFIELD(kem_info_4, krb5_kdc_kem_info, serverNonce, 4,
+         opt_ostring_data);
 static const struct atype_info *kdc_kem_info_fields[] = {
     &k5_atype_kem_info_0, &k5_atype_kem_info_1, &k5_atype_kem_info_2,
     &k5_atype_kem_info_3, &k5_atype_kem_info_4
